### ISSUE TITLE
WASM modules to skip PostScan Action and carry results back

### ISSUE THREAD

**Issue by varanasikalyan** [aquasecurity/trivy#3302]
Existing PostScan logic for WASM modules operates on a mandatory PostScan action.

https://github.com/aquasecurity/trivy/blob/01c7fb14bc6c09c41497509667d5455bf96b7e68/pkg/module/module.go#L494

Requesting to provide an alternate to skip the PostScan action and return the results,

https://github.com/aquasecurity/trivy/blob/01c7fb14bc6c09c41497509667d5455bf96b7e68/pkg/module/module.go#L489

Which can be used from the client end to operate on.

**Comment by varanasikalyan** [comment]
Adding the explanation for need for the above requests made

**Request 2**: DetectedVulnerability already has,

```
// Custom is for extensibility and not supposed to be used in OSS
Custom interface{} `json:",omitempty"`
```
But we noticed that this **Custom** attribute being **{}interface** type, using this field from inside .wasm file will require some strong type declaration for each vulnerability.

If one CVE(DetectedVulnerability) is detected under more than one .wasm module as vulnerable, **Custom** attribute being **{}interface** type will reset module names that we want to carry across.

Rather provide a strongly typed attribute like below which can be helpful.
```
// Contexts gets filled with all contextual scans detected WASM PostScan
Contexts []string `json:",omitempty"`

```
Please do suggest if this can be handled better?

**Request 3**: Currently trivy OSS picks .wasm files ".trivy/modules" folder, hardcoded, 

1. But this might not work for trivy scanner running in windows environment as **".trivy/modules"** is not Windows advisable folder structure. Hence we may need a flexibility to accept custom user input path.
2. We come across a situation where we many not require all the .wasm files under **".trivy/modules"**  to be registered by Trivy OSS, hence accept custom .wasm **file names/module names** only to be registered.

**Comment by knqyf263** [comment]
First of all, we should break down the issue into three issues.

>Request 1: Existing PostScan logic for WASM modules operates on a mandatory PostScan action.

I don't get it. Can you elaborate on it? If you don't implement `PostScan` in a module, the method will not be called.
https://github.com/aquasecurity/trivy/blob/5b944d20acbc238c0e41bfba7281acd20489843e/pkg/module/module.go#L385-L388

Isn't this what you need?

>Request 2: DetectedVulnerability already has,

Let me think about it.

>Request 3: Currently trivy OSS picks .wasm files ".trivy/modules" folder, hardcoded,

Does `--module-paths` meet your requirement? If yes, please create a dedicated issue.

**Comment by varanasikalyan** [comment]
@knqyf263, I've created a separate issue for **Request 3** https://github.com/aquasecurity/trivy/issues/3406
my mistake, apologies

**Comment by varanasikalyan** [comment]
@knqyf263 looks like we can work with **`Customs map[string]interface{}`** but changing this from, 
**Custom interface{} to Customs map[string]interface{}** looks like a breaking change for the existing trivy scan flows

We are already definitely using this attribute at few of trivy premium workflows.
I also see this attribute is already being used in all or most of os package detectors in trivy OSS.

and regarding the **Request 1**: is we are suggested to use mandatory Update - PostScan action, we may need to include having to append **`Customs map[string]interface{}`** under **updateResults** across various wasm file calls

https://github.com/aquasecurity/trivy/blob/b88bccae6e8619c5146a32c4d835d7a6a2be1e7c/pkg/module/module.go#L527

By making a change under **// Update vulnerabilities in the existing result 
`for _, got := range g.Vulnerabilities {
						if got.VulnerabilityID == v.VulnerabilityID && got.PkgName == v.PkgName &&
							got.PkgPath == v.PkgPath && got.InstalledVersion == v.InstalledVersion {`** 

for key, val := range got.Custom {
    v[key] = val
}

**Comment by knqyf263** [comment]
>I also see this attribute is already being used in all or most of os package detectors in trivy OSS.

Trivy OSS doesn't use `Custom`. It affects those who use `Custom`. If Aqua needs `Customs`, you just need to be aligned with that in other places that use `Custom`.

>append Customs map[string]interface{} under updateResults across various wasm file calls

Is it different from what I said here?
https://github.com/aquasecurity/trivy/issues/3302#issuecomment-1372161142

**Comment by varanasikalyan** [comment]
Hi @knqyf263, I completely agree about the flexibility of having to use an interface{} type which helps add any data tomorrow but we are using Custom at many places of code at Aqua end and will require an interpretation of the right datatype where ever `Customs map[string]interface{}` is being used due to this breaking change. This after consulting with other teammates we request if we can get

`Contexts map[string]interface{}` in addition to the existing `Custom interface{}` as you seem to be okay with the provision of Contexts from the above comment??

https://github.com/aquasecurity/trivy/issues/3302#issuecomment-1372161142

Please do the needful.

**Comment by knqyf263** [comment]
>but we are using Custom at many places of code at Aqua end and will require an interpretation of the right datatype where ever Customs map[string]interface{} is being used due to this breaking change.

I don't think this is a breaking change. It is used only by Aqua. You can think it is part of the Aqua source code. It is kind of like internal refactoring. I think you are too afraid of refactoring. How hard is it to change to `Customs`? If we decide to add new fields forever, it will be technical debt.

**Comment by varanasikalyan** [comment]
Hi @knqyf263, we have made an attempt to check if Custom interface{} to Customs map[string]interface{} would work. We observed 
1. panic errors with the easyjson marshalling / unmarshalling with above types Custom interface{}, Customs map[string]interface{} especially when communicating with running WASM module.
2. We also observed a panic statement `panic: unimplemented: (reflect.Type).Key()`

From understanding where the issue is coming in from, we found,

https://github.com/mailru/easyjson/issues/121
https://github.com/tinygo-org/tinygo/pull/2734

Please suggest how we can move forward from here.

**Comment by knqyf263** [comment]
We should create a small PoC and test it. There are several steps.

1. Try easyjson marshaling structs with interface in Go
2. Try easyjson marshaling structs with interface in TinyGo
3. Try easyjson marshaling structs with interface in Wasm + TinyGo

Did you try the first one?

**Comment by varanasikalyan** [comment]
Hi @knqyf263 , we have attempted 1st step "Try easyjson marshaling structs with interface in Go" and they are working okay.
We request you to try the next 2 options to see if usage of interface{} types with WASM modules usage is feasible or not. Else, please help us provide with further feasible options.

**Comment by knqyf263** [comment]
> We request you to try the next 2 options

You can just use TinyGo instead of Go.

**Comment by varanasikalyan** [comment]
Hi @knqyf263, requesting you to please help us here, As we understand the use of TinyGo is for microcontrollers or WASMs and we have already used TinyGo to build and generate WASM file to see if its working as expected. But we observed its failing with map[string]interface{}

We request you to try the TinyGo approach for POC as we are also added up in other priority items already.

Please do the needful.

**Comment by knqyf263** [comment]
>But we observed its failing with map[string]interface{}

Does it mean you tried to build `map[string]interface{}` with TinyGo and it failed?

**Comment by knqyf263** [comment]
No, I'm asking about a small PoC. Trivy is complicated. We need to break down the issue.
